SU2_CFD/src/solvers/CSpeciesSolver.cpp

```diff
@@ -400,37 +395,20 @@ void CSpeciesSolver::BC_Inlet(CGeometry* geometry, CSolver** solver_container, C
       if (dynamic_grid)
         conv_numerics->SetGridVel(geometry->nodes->GetGridVel(iPoint), geometry->nodes->GetGridVel(iPoint));
 
-      if(bounded_scalar){
-        /*--- Obtain flow velocity vector at inlet boundary node ---*/
-        su2double Velocity_Inlet[MAXNDIM] = {0.0}, density_inlet;
-        for (auto iDim = 0u; iDim < nDim; iDim++) Velocity_Inlet[iDim] = V_inlet[iDim + 1];//solver_container[FLOW_SOL]->GetNodes()->GetVelocity(iPoint, iDim);
-        
-        /*--- Obtain density at inlet boundary node ---*/
-        density_inlet = solver_container[FLOW_SOL]->GetNodes()->GetDensity(iPoint);
-
-        /*--- Compute mass flux on current node ---*/
-        EdgeMassFlux = density_inlet * GeometryToolbox::DotProduct(nDim, Velocity_Inlet, Normal);
-      
-        /*--- Communicate inlet mass flux to numerics ---*/
-        conv_numerics->SetMassFlux(EdgeMassFlux);
+      if (conv_numerics->GetBoundedScalar()) {
+        const su2double* velocity = &V_inlet[prim_idx.Velocity()];
+        const su2double density = solver_container[FLOW_SOL]->GetNodes()->GetDensity(iPoint);
+        conv_numerics->SetMassFlux(BoundedScalarBCFlux(iPoint, implicit, density, velocity, Normal));
       }
 
       /*--- Compute the residual using an upwind scheme ---*/
+
       auto residual = conv_numerics->ComputeResidual(config);
       LinSysRes.AddBlock(iPoint, residual);
 
       /*--- Jacobian contribution for implicit integration ---*/
-      const bool implicit = (config->GetKind_TimeIntScheme() == EULER_IMPLICIT);
-      if (implicit) Jacobian.AddBlock2Diag(iPoint, residual.jacobian_i);
-
-      /*--- Applying convective flux correction to negate the effects of flow divergence ---*/
-      if(bounded_scalar){
-        for(unsigned short iVar=0; iVar<nVar; iVar++){
-          su2double FluxCorrection_i = GetNodes()->GetSolution(iPoint, iVar) * EdgeMassFlux;
 
-          LinSysRes(iPoint, iVar) -= FluxCorrection_i;
-        }
-      }if (implicit) Jacobian.AddVal2Diag(iPoint, max(0.0, EdgeMassFlux));
```

Comment:
This is what is meant by not duplicating code. And this is the last time I do this, going forward if I don't see any effort in integrating things properly I will simply request changes.